p2p/src/network/yamux/p2p_network_yamux_reducer.rs

```diff
@@ -318,21 +349,45 @@ impl P2pNetworkYamuxState {
                             });
                         }
                     }
+                    YamuxFrameInner::WindowUpdate { .. } => {
+                        while let Some(frame) = pending_outgoing.pop_front() {
+                            dispatcher.push(P2pNetworkYamuxAction::OutgoingFrame { addr, frame });
+                        }
+                    }
                     _ => {}
                 }
 
                 Ok(())
             }
-            P2pNetworkYamuxAction::OutgoingFrame { frame, addr } => {
+            P2pNetworkYamuxAction::OutgoingFrame { mut frame, addr } => {
                 let Some(stream) = yamux_state.streams.get_mut(&frame.stream_id) else {
                     return Ok(());
                 };
-                match &frame.inner {
+                match &mut frame.inner {
                     YamuxFrameInner::Data(data) => {
-                        // must not underflow
-                        // the action must not dispatch if it doesn't fit in the window
-                        // TODO: add pending queue, where frames will wait for window increase
-                        stream.window_theirs = stream.window_theirs.wrapping_sub(data.len() as u32);
+                        if let Some(new_window) =
+                            stream.window_theirs.checked_sub(data.len() as u32)
+                        {
+                            // their window is big enough, decrease the size
+                            // and send the whole frame
+                            stream.window_theirs = new_window;
+                        } else if stream.window_theirs != 0 && stream.pending.is_empty() {
+                            // their window is not big enough, but has some space,
+                            // and the queue is empty,
+                            // do not send the whole frame,
+                            // split it and put remaining in the queue,
+                            if let Some(remaining) = frame.split_at(stream.window_theirs as usize) {
+                                stream.pending.push_back(remaining);
+                            }
+                            // the window will be zero after sending
+                            stream.window_theirs = 0;
+                        } else {
+                            // either the window cannot accept any byte,
+                            // or the queue is already not empty
+                            // in both cases the whole frame goes in the queue and nothing to send
+                            stream.pending.push_back(frame);
```

Comment:
Need to add a limit here. Malicious peer can keep its window small and send an RPC request in a loop.